contents/blog/null-handling-hogql.md

```diff
@@ -0,0 +1,82 @@
+
+To solve the issue with nulls, Marius and Michael came up with a bunch of potential solutions, (none of which were inevitably chosen):
+
+1. **Returning a default value depending on type.** For example, returning `""` for strings, `0` for numbers, or `false` for booleans. This didn’t handle the expression `properties.bla is null` and required rewriting `is set` and `is not set` logic.
+
+2. **Including a nullable checkbox in data management.** Similar to the previous option, but required users to input if they expect a property to be null. This required a lot of user work and remains confusing if they didn’t understand how to handle null.
+
+3. **Create JS-like `'bla' (NOT) IN properties` expression logic.** We felt this was too complex for users and makes the question of `is null` for an unset property unclear. 
+
+4. **Use the ClickHouse function `sumOrDefaultIf`.** Faces the same problem of being null if any values are null.
+
+5. **Use non-ANSI compliant string operator.** ANSI is SQL standards for function and syntax. For example, `f'{NULL}-haha-{properties.bla}'` is non-ANSI compliant because it doesn’t conform to standard SQL string concatenation syntax. We decided it is better to keep the standard operators and be flexible with types instead.
+
+## The solution(s) to our null problems
+
+The solution was to handle null for values we expect to be strings differently than numbers.
+
+For numbers, returning `0` for null number properties meaningfully affects calculations, because they get added as non-null to counts. To prevent this, PostHog continues to return null and lets ClickHouse skip these in aggregations.
+
+Strings don’t need the same accuracy, users mostly want to return or concat strings anyway. To solve the specific issue with the concat function, Marius made it null-tolerant. This means every value converts to a string. For example, `concat(null, 'a', 3, toString(4), toString(NULL), properties.$screen_width)` turns into `concat('', 'a', toString(3), toString(4), '')`. 
+
+To do this, PostHog automatically adds `ifNull` falling back to empty strings (`''`) to concat functions. HogQL abstracts this away because the functionality creates a long and "dorky" ClickHouse SQL strings like:
+
+```
+-- hogql
+concat(properties.$screen_width, 'x', properties.$screen_height)
+
+-- clickhouse
+concat(ifNull(toString(replaceRegexpAll(nullIf(nullIf(JSONExtractRaw(events.properties, '$screen_width', ''), 'null'), '^\"|\"$', '')), ''), 'x', ifNull(toString(replaceRegexpAll(nullIf(nullIf(JSONExtractRaw(events.properties, '$screen_height'), ''), 'null'), '^\"|\"$', '')), ''))
+```
+
+By understanding the common use cases and objectives of users when interacting with nulls, Marius implemented these solutions to streamline the user experience. He also didn’t implement a solution moving us too far away from standard practice, making HogQL more maintainable for us and understandable for users.
+
+## Further reading
```

Comment:
Feels like a bit of an abrupt ending. From the title, I was expecting more scenarios and edge cases. Perhaps from the term "adventures", I was expecting more.

Can we add more to this post? Or maybe rename the title so that it sets expectations appropriately?

e.g. "Why null handling in SQL and Clickhouse is super annoying"